New Item Extension mod element

Goldorion opened the pull request:
This PR adds the new Item Extension mod element as we discussed a few times on the Discord server.
![image](https://user-images.githubusercontent.com/38427877/158083367-24c74048-1449-4724-973d-be72117306ec.png)

I did not keep the Fuelmaker panel because it would mean adding an MCItemHolder for each property type. I also think this makes the UI better.

Changelog:
- Added a new Item Extension class
- Fuel mod element has been moved to this ME
- Added a new success condition for the fuel property
- Moved the custom dispenser property to this ME
- Added a new composter property in this ME
- Added 2 converts for older workspaces (1 for fuels, 1 for items)

I kept the Fuel class, so the converter is much cleaner. I also kept dispenser property variables inside the Item for the same reason.


@Defeatomizer I will need the new Russian tips.

Defeatomizer commented on plugins/generator-1.16.5/forge-1.16.5/templates/itemextension.java.ftl:
```diff
@@ -39,7 +39,7 @@ package ${package}.item.extension;
 		@SubscribeEvent
 		public void init(FMLCommonSetupEvent event) {
-		    <#if data.isCompostable>
+		    <#if (extension.layerChance > 0)>
```
```suggestion
		    <#if data.layerChance gt 0>
```

Defeatomizer commented on plugins/mcreator-localization/help/default/item_extension/enable_fuel.md:
```diff
@@ -0,0 +1 @@
+Check this case to make the selected item a fuel for furnaces like coal.
```
Coal is not a furnace.
```suggestion
Check this box to make the selected item a fuel for furnaces (like coal, planks, etc.).
```

Defeatomizer commented on plugins/mcreator-localization/help/default/item_extension/item.md:
```diff
@@ -0,0 +1 @@
+This is the item that will receive the properties.
```
I don't think properties can be received (at least never heard that).
```suggestion
This is the target item of this extension.
```

Defeatomizer commented on plugins/mcreator-localization/help/default/item_extension/layer_chance.md:
```diff
@@ -0,0 +1,4 @@
+This number is the percentage of chance of the item to add a layer to the composter block.
+Keep the value to 0 to disable it.
```
```suggestion
Keep the value of 0 to disable it.
```

Defeatomizer commented on plugins/mcreator-localization/lang/texts.properties:
```diff
@@ -1216,6 +1217,7 @@ modelement.food.description=Food is an item that can be eaten and can give you p
 modelement.fuel.description=Fuel is a block or item that can be used as fuel in furnaces.
 modelement.function.description=Function allows players to run command sets when called.
 modelement.gui.description=GUI is a displayable panel with elements such as buttons, labels, slots, ...
+modelement.itemextension.description=An Item extension is an extra property that can be added to an item.
```
Other mod elements don't have articles (a/an/the) in their descriptions.
```suggestion
modelement.itemextension.description=Item extension is a set of extra properties that can be added to an item.
```

Defeatomizer commented on plugins/mcreator-localization/lang/texts.properties:
```diff
@@ -1940,6 +1941,16 @@ elementgui.gui.import_overlay_base_texture=Click this to import overlay base tex
+elementgui.item_extension.fuel_properties=Custom fuel properties
+elementgui.item_extension.enable_fuel=Make this item a fuel:
+elementgui.item_extension.burn_time=Fuel burn time (power):
+elementgui.item_extension.layer_chance=Chance of success to compost
```
To follow your other strings:
```suggestion
elementgui.item_extension.layer_chance=Chance of success to compost:
```

Defeatomizer commented on src/main/java/net/mcreator/element/ModElementType.java:
```diff
@@ -164,4 +165,7 @@ public boolean hasProcedureTriggers() {
 	public static ModElementType<?> TAG;
 	public static ModElementType<?> TOOL;
 
+
+	// Legacy
```
Isn't one empty line enough?
```suggestion
	// Legacy
```

Defeatomizer commented on src/main/java/net/mcreator/ui/modgui/ItemExtensionGUI.java:
```diff
@@ -0,0 +1,168 @@
+				L10N.t("elementgui.item_extension.fuel_properties"), TitledBorder.LEADING,
+				TitledBorder.DEFAULT_POSITION, getFont(), (Color) UIManager.get("MCreatorLAF.BRIGHT_COLOR")));
+
+		// Composter
```
Composter is a block, compostability is an item property.
```suggestion
		// Compostability
```

Defeatomizer replied:
So what about these 3 suggestions?

Defeatomizer commented on src/main/java/net/mcreator/ui/modgui/ItemGUI.java:
```diff
@@ -474,11 +434,11 @@ else if (page == 0)
-		dispenseResultItemstack.setSelectedProcedure(item.dispenseResultItemstack);
-
-		updateDispenseElements();
+//		hasDispenseBehavior.setSelected(item.hasDispenseBehavior);
+//		dispenseSuccessCondition.setSelectedProcedure(item.dispenseSuccessCondition);
+//		dispenseResultItemstack.setSelectedProcedure(item.dispenseResultItemstack);
+//
+//		updateDispenseElements();
```
Why do you keep these lines?
```suggestion
```

Goldorion replied:
I was doing some tests and then, I forgot those comments.

Defeatomizer commented on src/main/java/net/mcreator/ui/modgui/ItemGUI.java:
```diff
@@ -517,9 +477,9 @@ else if (page == 0)
-		item.hasDispenseBehavior = hasDispenseBehavior.isSelected();
-		item.dispenseSuccessCondition = dispenseSuccessCondition.getSelectedProcedure();
-		item.dispenseResultItemstack = dispenseResultItemstack.getSelectedProcedure();
+//		item.hasDispenseBehavior = hasDispenseBehavior.isSelected();
+//		item.dispenseSuccessCondition = dispenseSuccessCondition.getSelectedProcedure();
+//		item.dispenseResultItemstack = dispenseResultItemstack.getSelectedProcedure();
```
Why do you keep these lines?
```suggestion
```

Defeatomizer reviewed (commented):
I forgot I'm not a maintainer...
Maybe there's a way to make these new converters cooperate with each other to reduce the number of item extensions added (i. e. if there's a fuel mod element that is applied to a custom item with dispense behavior, it results into 1 item extension instead of 2).
Also note that this ticket resolves #56.

Defeatomizer commented:
In case accepted, these suggestions should go in a patch I guess...
Re-requesting myself for the tips translations.

KlemenDEV commented:
@Defeatomizer you can open PR to his PR, it is much better and standard this way

Defeatomizer commented on src/test/java/net/mcreator/integration/TestWorkspaceDataProvider.java:
```diff
@@ -1144,6 +1136,19 @@ private static GeneratableElement getExampleFor(ModElement modElement, Random ra
+
+			itemExtension.enableFuel = !emptyLists;
+			itemExtension.fuelPower = new int[] { 0, 100, 1000, 4000 }[valueIndex];
+			itemExtension.isCompostable = emptyLists && _true;
```
```suggestion
```

Goldorion replied:
I totally forgot to check tests when i removed it 😅

Defeatomizer reviewed (changes requested):
I found the reason of failing tests.

Defeatomizer commented:
OKay, I didn't check the entire PR :man_facepalming:
>    2022-03-13-10:14:49 [Test worker/ERROR] [Template Generator] Failed to generate template: itemextension.java.ftl
    freemarker.core.InvalidReferenceException: The following has evaluated to null or missing:
    ==> data.isCompostable  [in template "itemextension.java.ftl" at line 39, column 42]

Defeatomizer commented on plugins/generator-1.16.5/forge-1.16.5/templates/itemextension.java.ftl:
```diff
@@ -0,0 +1,88 @@
+
+@Mod.EventBusSubscriber(bus = Mod.EventBusSubscriber.Bus.MOD) public class ${name}ItemExtension {
+
+	<#if data.hasDispenseBehavior || data.isCompostable>
```
```suggestion
	<#if data.hasDispenseBehavior || data.layerChance gt 0>
```

Defeatomizer commented:
> @Defeatomizer you can open PR to his PR, it is much better and standard this way

Oops, too late.

MagicalAlexey commented:
throw off the Discord please

Max094Reikeb commented:
> give Discord please

???

MagicalAlexey commented:
> > give Discord please
> 
> ???

Discord server

Max094Reikeb commented:
> > > give Discord please
> > 
> > 
> > ???
> 
> Discord server

You don't make sense

Goldorion commented:
> give Discord please

This is a private Discord server for the team members only, so no.

Goldorion commented:
I just added a new success condition for the fuel property.

I will also need the Russian tips @Defeatomizer if you don't open another PR later.

Goldorion commented:
> Maybe there's a way to make these new converters cooperate with each other to reduce the number of item extensions added (i. e. if there's a fuel mod element that is applied to a custom item with dispense behavior, it results into 1 item extension instead of 2).

From what I know, the converter system is very limited and it doesn't allow us this. When the workspace opens, it iterates all mod elements and when a mod element can get a converter (its FV< FV of the latest converter for this MET), the converter(s) will be applied. Then, the next mod element is tested. If there is a way to do what you explained, it would require another special converter or another converter system, but that's out of the scope of this PR.

Goldorion commented on plugins/generator-1.16.5/forge-1.16.5/templates/itemextension.java.ftl:
```diff
@@ -0,0 +1,90 @@
+
+                        <#if hasProcedure(data.dispenseResultItemstack)>
+                            boolean success = this.isSuccessful();
+                            <#if hasReturnValueOf(data.dispenseResultItemstack, "logic")>
```
We need a boolean value for the condition, so why do we need to check for `itemstack`?

AlsoSomeoneElse replied:
Because the "Dispense result itemstack" can return an itemstack, not a boolean. Currently the returned itemstack is always ignored

Goldorion replied:
Oh sorry, I didn't read correctly the procedure. I thought we were with the fuel condition. Sorry 😅 You're right.

Goldorion commented on plugins/generator-1.16.5/forge-1.16.5/templates/itemextension.java.ftl:
```diff
@@ -0,0 +1,90 @@
+
+                        <#if hasProcedure(data.dispenseResultItemstack)>
+                            boolean success = this.isSuccessful();
+                            <#if hasReturnValueOf(data.dispenseResultItemstack, "logic")>
```
```suggestion
                            <#if hasReturnValueOf(data.dispenseResultItemstack, "itemstack")>
```

Goldorion commented on plugins/generator-1.18.2/forge-1.18.2/templates/itemextension.java.ftl:
```diff
@@ -0,0 +1,67 @@
+
+    				<#if hasProcedure(data.dispenseResultItemstack)>
+    					boolean success = this.isSuccess();
+    					<#if hasReturnValueOf(data.dispenseResultItemstack, "logic")>
```
```suggestion
    					<#if hasReturnValueOf(data.dispenseResultItemstack, "itemstack")>

```

Defeatomizer commented on plugins/mcreator-localization/help/default/item_extension/enable_fuel.md:
```diff
@@ -0,0 +1 @@
+Check this box to make the selected item a fuel for furnaces (like coal, planks, etc.).
```
Определяет, можно ли использовать выбранный предмет как топливо в печках (подобно углю, доскам и т. д.).

Goldorion commented:
> Oh cool xD

I didn't want to leave BOTW on my main screen to open Intellij, so I took the other way 😂

Defeatomizer commented on plugins/mcreator-localization/help/default/item_extension/fuel_success_condition.md:
```diff
@@ -0,0 +1 @@
+The condition has to be valid to use the selected item as fuel. If the condition fails, the item will not work as fuel until the condition successes.
```
Если условие указано, выбранный предмет может являться топливом только при выполнении этого условия.
```suggestion
If condition is specified, the selected item can only be used as fuel if the condition passes.
```

Defeatomizer commented on plugins/mcreator-localization/help/default/item_extension/item.md:
```diff
@@ -0,0 +1 @@
+This is the target item of this extension.
```
Определяет, какой предмет будет обладать данными свойствами.

Defeatomizer commented on plugins/mcreator-localization/help/default/item_extension/layer_chance.md:
```diff
@@ -0,0 +1,4 @@
+This number is the percentage of chance of the item to add a layer to the composter block.
```
Определяет вероятность (в процентах) превращения заданного предмета в перегной при использовании на компостнице.
Укажите 0 для отключения этой опции.

Список стандартных значений можно найти \[здесь](https://minecraft.fandom.com/wiki/Composter#Composting).

Defeatomizer reviewed (changes requested):
The help tip with the suggestion is translation of the in-app tip with that suggestion.

KlemenDEV commented:
Compost section should have a titled border too

KlemenDEV commented:
I think #1557 should be part of this new mod element instead

Goldorion commented:
> I think #1557 should be part of this new mod element instead

I think we can do this. However, is it possible to implement it in another PR. Max already has all the code ready and it would take me more time to understand, re-implement and test. Something we can do, if you want villager trades at the same time as the villager trade feature, is finish this PR first and then, merge it in a temp branch, so Max can adapt its PR.

AlsoSomeoneElse commented:
> I think #1557 should be part of this new mod element instead

I disagree. I think villager trades are different enough from item extensions, in that they take one or two items and return one item, while IE take one item and perform some actions

Goldorion commented:
> Compost section should have a titled border too

Done

Defeatomizer commented:
[Compost](https://github.com/MCreator/MCreator/pull/2479/commits/4f4b9279886949dabd708b6bb71f68f599a69e23#diff-1589a2170fb1acc56a231e7292293d2ab4d00098bc2e3b145975bf4e3fdb24de) is what the target item turns into... :man_facepalming:

Goldorion commented:
> [Compost](https://github.com/MCreator/MCreator/pull/2479/commits/4f4b9279886949dabd708b6bb71f68f599a69e23#diff-1589a2170fb1acc56a231e7292293d2ab4d00098bc2e3b145975bf4e3fdb24de) is what the target item turns into... 🤦‍♂️

So the title is appropriated...

AlsoSomeoneElse commented on plugins/generator-1.18.2/forge-1.18.2/templates/elementinits/fuels.java.ftl:
```diff
@@ -42,9 +43,14 @@ package ${package}.init;
+		<#list itemextensions as extension>
+            <#if extension.enableFuel>
+                if (event.getItemStack().getItem() == ${mappedMCItemToItem(extension.item)})
+                    <#if hasReturnValueOf(extension.fuelSuccessCondition, "logic")>if (<@procedureOBJToConditionCode extension.fuelSuccessCondition/>)</#if>
```
`hasReturnValueOf` can be replaced with `hasProcedure` because the boolean value isn't optional (same for the 1.16 template)

AlsoSomeoneElse commented on plugins/generator-1.18.2/forge-1.18.2/itemextension.definition.yaml:
```diff
@@ -0,0 +1,15 @@
+templates:
+  - template: itemextension.java.ftl
```
This template and class could be renamed to something like [name]DispenseBehavior, and should be deleted when the condition is false

Goldorion replied:
Currently, it only contains the code for the dispenser behaviour features, but in the future, we may add more features to it.

AlsoSomeoneElse replied:
I think it'd be cleaner to keep unrelated behaviors in different classes, similar to what's happening with composting and dispense behavior

AlsoSomeoneElse commented on src/main/java/net/mcreator/ui/modgui/ItemExtensionGUI.java:
```diff
@@ -0,0 +1,191 @@
+	private final JCheckBox enableFuel = L10N.checkbox("elementgui.common.enable");
+	private final JSpinner fuelPower = new JSpinner(new SpinnerNumberModel(1600, 0, Integer.MAX_VALUE, 1));
+	// Compostable
+	private final JSpinner layerChance = new JSpinner(new SpinnerNumberModel(0, 0.00f, 1.00f, 0.05f));
```
I'd replace the spinner model to `(0, 0, 1, 0.01)` so that it uses doubles and allows more precise values

Goldorion replied:
Users should not use a double value because Minecraft uses a float variable for the chance. I am only using a double to save it because the number spinner object uses a double.

AlsoSomeoneElse replied:
I tested with the current values and the spinner doesn't work properly (if it's at 0.05, it wouldn't go down to 0, same for 0.95 to 1). Moving to double did fix the issue

AlsoSomeoneElse commented on src/main/java/net/mcreator/element/types/ItemExtension.java:
```diff
@@ -0,0 +1,53 @@
+	public ItemExtension(ModElement element) {
+		super(element);
+		fuelPower = 1600;
+		layerChance = 0.65f;
```
This could be 0.65 since it's a double

AlsoSomeoneElse commented on plugins/generator-1.16.5/forge-1.16.5/templates/itemextension.java.ftl:
```diff
@@ -0,0 +1,90 @@
+            public static void furnaceFuelBurnTimeEvent(FurnaceFuelBurnTimeEvent event) {
+		        ItemStack itemstack = event.getItemStack();
+                if(event.getItemStack().getItem() == ${mappedMCItemToItem(data.item)})
+                    <#if hasReturnValueOf(data.fuelSuccessCondition, "logic")>if (<@procedureOBJToConditionCode data.fuelSuccessCondition/>)</#if>
```
See comment on 1.18 template

AlsoSomeoneElse replied:
```suggestion
                    <#if hasProcedure(data.fuelSuccessCondition)>if (<@procedureOBJToConditionCode data.fuelSuccessCondition/>)</#if>
```

AlsoSomeoneElse commented:
One last suggestion, I think it would be useful to use a fixed value/procedure selector for the fuel duration, similar to the one used by blocks for emitted redstone power

Goldorion commented:
> One last suggestion, I think it would be useful to use a fixed value/procedure selector for the fuel duration, similar to the one used by blocks for emitted redstone power

I can try to change the current thing for this. However, I'm not sure about being able to make a good converter to create a procedure. I'll try, but I can't promise. 😅

Goldorion commented:
> One last suggestion, I think it would be useful to use a fixed value/procedure selector for the fuel duration, similar to the one used by blocks for emitted redstone power

I will just need some help to make the JSpinner bigger... 
![image](https://user-images.githubusercontent.com/38427877/158492822-027032b2-0751-4f99-92ed-678f810cca53.png)

Goldorion commented:
When I tested again the converter, I noticed to make it works, I need to keep registering the Fuel ME as other MEs. However, it makes it appear everywhere, so I simply disabled the code for the Fuel ME when it's called (using looping all MEs). It also means @Defeatomizer icons need to stay.
@AlsoSomeoneElse I also changed the basic JSpinner to a NumberProcedureSelector. As the condition, we can only use the `itemstack` dependency.

Defeatomizer reviewed (changes requested):
English dispense behavior tips somehow are in the Russian folder, please rearrange them.

Defeatomizer reviewed (changes requested):
You should also remove the old Russian tips, not just recreate them in another directory.

Goldorion commented:
> You should also remove the old Russian tips, not just recreate them in another directory.

I deleted them in a previous commit... :/

Defeatomizer commented:
[There](https://github.com/MCreator/MCreator/tree/6f450f8a7ebab488f89d4cce73c4d333bc7fee3b/plugins/mcreator-localization/help/ru_RU) I can still see fuel directory...

Defeatomizer reviewed (changes requested):
>Seems like the encoding of Russian tips have changed, could you please revert that?

Goldorion commented:
> > Seems like the encoding of Russian tips have changed, could you please revert that?

Ok, but revert what?

Defeatomizer commented on plugins/mcreator-localization/help/ru_RU/item_extension/burn_time.md:
```diff
@@ -8,4 +8,4 @@
 
 Кликните [здесь](https://mcreator.net/wiki/burn-time-fuels), чтобы узнать время сгорания некоторых "ванильных"
-предметов.
+предметов.
```
Change LS back, please.

Goldorion replied:
It's already the right one. :/

Defeatomizer replied:
What happens if you remove all the text here (Ctrl + A) and copy the text (Ctrl + A) of the old tip?

Defeatomizer commented on plugins/mcreator-localization/help/ru_RU/item_extension/dispense_result_itemstack.md:
```diff
@@ -1,4 +1,4 @@
-﻿Эта процедура определяет, что произойдёт с предметом при попытке его раздачи. Если ничего не выбрано либо указанная
-процедура не возвращает предмет, количество предметов в изначальной стопке в случае успешной раздачи уменьшится на 1.
-
+Эта процедура определяет, что произойдёт с предметом при попытке его раздачи. Если ничего не выбрано либо указанная
+процедура не возвращает предмет, количество предметов в изначальной стопке в случае успешной раздачи уменьшится на 1.
+
```
Change LS back, please.

Defeatomizer commented on plugins/mcreator-localization/help/ru_RU/item_extension/has_dispense_behavior.md:
```diff
@@ -1,2 +1,2 @@
-Если флажок установлен, раздатчик не выбросит данный предмет, а вызовет процедуры
-"${l10n.t("elementgui.item.dispense_success_condition")}" и "${l10n.t("elementgui.item.dispense_result_itemstack")}".
+Если флажок установлен, раздатчик не выбросит данный предмет, а вызовет процедуры
+"${l10n.t("elementgui.item.dispense_success_condition")}" и "${l10n.t("elementgui.item.dispense_result_itemstack")}".
```
Change LS back, please.

Defeatomizer commented on plugins/mcreator-localization/help/ru_RU/item_extension/dispense_success_condition.md:
```diff
@@ -0,0 +1,4 @@
+Задаёт условие успешной раздачи. Если оно не удовлетворяется, раздачи не произойдёт.
```
Change LS back, please.

Defeatomizer commented on src/main/java/net/mcreator/element/converter/fv28/FuelToItemExtensionConverter.java:
```diff
@@ -0,0 +1,98 @@
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+/*
+ * MCreator (https://mcreator.net/)
+ * Copyright (C) 2012-2020, Pylo
+ * Copyright (C) 2020-2022, Pylo, opensource contributors
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+/*
+ * MCreator (https://mcreator.net/)
+ * Copyright (C) 2012-2020, Pylo
+ * Copyright (C) 2020-2022, Pylo, opensource contributors
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
```
Intellij is not stupid for lots of other reasons.
```suggestion
```

Goldorion replied:
I noticed every time I move a class with the refractor feature, it adds another header.

Defeatomizer replied:
You can use DND instead.

Goldorion replied:
What is DND?

Defeatomizer replied:
[Drag aNd Drop](https://github.com/MCreator/MCreator/blob/55a42ee2f140e392022efb17be8b4663b3371dd8/src/main/java/net/mcreator/ui/MCreatorDropTarget.java)

KlemenDEV commented on src/main/java/net/mcreator/element/types/Item.java:
```diff
@@ -80,6 +80,7 @@
 	public Procedure onFinishUsingItem;
 
+	// Legacy
 	public boolean hasDispenseBehavior;
```
If this is no longer used, it should be removed, not marked as legacy

KlemenDEV commented on src/test/java/net/mcreator/integration/TestWorkspaceDataProvider.java:
```diff
@@ -89,9 +89,10 @@ public static List<GeneratableElement> getModElementExamplesFor(Workspace worksp
 			generatableElements.add(getToolExample(me(workspace, type, "11"), "Shears", random, true, false));
 			generatableElements.add(getToolExample(me(workspace, type, "12"), "Fishing rod", random, true, false));
-		} else if (type == ModElementType.FUEL || type == ModElementType.TAB) {
+		} else if (type == ModElementType.ITEMEXTENSION || type == ModElementType.TAB) {
```
This section that adss only 3 tests is for really simple elements that ITEMEXTENSIO, so move ITEMEXTENSION to standard tests

Goldorion commented:
Sorry for the requested revikew. tests appeared in green for me, but then, they changed to red. 🤔

KlemenDEV commented on src/test/java/net/mcreator/integration/TestWorkspaceDataProvider.java:
```diff
@@ -1127,6 +1119,19 @@ null, null, new MItemBlock(modElement.getWorkspace(),
+			itemExtension.item = new MItemBlock(modElement.getWorkspace(), getRandomMCItem(random, blocksAndItems).getName());
+
+			itemExtension.enableFuel = !emptyLists;
+			itemExtension.fuelPower = new NumberProcedure(emptyLists ? null : "number1", 1600);
```
This is not ok. !emptyLists condition fuel generation, then emptyLists conditions procedures, so procedures are never tested

KlemenDEV commented on src/main/java/net/mcreator/element/types/ItemExtension.java:
```diff
@@ -20,22 +21,32 @@
 
 	@Override public BufferedImage generateModElementPicture() {
-		return MinecraftImageGenerator.Preview.generateFuelPreviewPicture(getModElement().getWorkspace(), block);
+		return MinecraftImageGenerator.Preview.generateFuelPreviewPicture(getModElement().getWorkspace(), item);
```
generateFuelPreviewPicture? Don't think this is relavant, so cleanup this code

Goldorion replied:
I kept this preview because I didn't know what could represent the entire element. I'm open to ideas.

KlemenDEV replied:
We can use mod element icon, no need to generate preview picture

KlemenDEV replied:
You don't have to override for default icon, as this is already handled in parent class

Goldorion replied:
Nothing is overridden...

KlemenDEV replied:
`@Override public BufferedImage generateModElementPicture() {`

to me it seems like this is overriding default generateModElementPicture

Goldorion replied:
![image](https://user-images.githubusercontent.com/38427877/169320187-141a018b-fe01-4aa3-9d49-f8fd70a9cbdc.png)
to me it seems like there is no such method

KlemenDEV replied:
Hmm, maybe github cached code view for me

AlsoSomeoneElse commented on plugins/generator-1.18.2/forge-1.18.2/templates/elementinits/fuels.java.ftl:
```diff
@@ -42,9 +43,17 @@ package ${package}.init;
+		ItemStack itemstack = event.getItemStack();
+		<#list itemextensions as extension>
+            <#if extension.enableFuel>
+                if (event.getItemStack().getItem() == ${mappedMCItemToItem(extension.item)})
```
Here you could replace `event.getItemStack()` with `itemstack`

AlsoSomeoneElse commented:
The "Dispense successfully" tooltips still refer to `elementgui.item.dispense_result_itemstack`

Goldorion commented:
> The element converter would attempt to turn an item into an item extension, or return `null`, both of which shouldn't happen

Why the item should not be returned as an item extension if it successes?

AlsoSomeoneElse commented:
> > The element converter would attempt to turn an item into an item extension, or return `null`, both of which shouldn't happen
> 
> Why the item should not be returned as an item extension if it successes?

Because the converter should return the original element after the conversion. By returning the item extension, you're basically saying "This item is now an item extension"

Goldorion commented:
> > > The element converter would attempt to turn an item into an item extension, or return `null`, both of which shouldn't happen
> > 
> > 
> > Why the item should not be returned as an item extension if it successes?
> 
> Because the converter should return the original element after the conversion. By returning the item extension, you're basically saying "This item is now an item extension"

I based this converter on the FoodToItem converter as Klemen approved it. So fv28 should also be fixed as it has [the same problem](https://github.com/MCreator/MCreator/blob/fed67fd924913e3b5f476385a0cde377b5433038/src/main/java/net/mcreator/element/converter/fv28/FoodToItemConverter.java#L103).

AlsoSomeoneElse commented:
If possible, you should try to make the GUI larger. As of now, the fuel power value is cut after 3 digits
![fuel value](https://user-images.githubusercontent.com/64546452/170788725-09a43ffb-442c-44b2-9dfa-1e83a848e1d2.png)

Goldorion commented:
> If possible, you should try to make the GUI larger. As of now, the fuel power value is cut after 3 digits

I can't resize this single Spinner as it is a part of `NumberProcedureSelector`. So what I made is simply made the JSpinner for all NumberProcedureSelector bigger. It can now properly display 6 digits. 
![image](https://user-images.githubusercontent.com/38427877/170799979-796c51df-ed37-495c-a6ef-a416bd26f683.png)

Defeatomizer commented:
I'd recommend integrating the burn time label and help tip into this selector (this should also give you some Mode free space for the spinner inside it).

Defeatomizer commented:
I'd recommend integrating the burn time label and help tip into this selector (this should also give you some more free space for the spinner inside it).

KlemenDEV commented:
Spinner size should change based on the container you put the numberprocselector into, so you shouldn't need to change it.

If you do, open new PR for it and make new constructor with width parameter or setter method so other UIs are not changed / potentially broken

Goldorion commented:
> I'd recommend integrating the burn time label and help tip into this selector (this should also give you some more free space for the spinner inside it).

This is the first thing I tried and it didn't change the size of the JSpinner.

Goldorion commented:
> Spinner size should change based on the container you put the numberprocselector into, so you shouldn't need to change it.
> 
> If you do, open new PR for it and make new constructor with width parameter or setter method so other UIs are not changed / potentially broken

I tried a lot of things since the beginning and the thing I've done is the only thing that works...

KlemenDEV commented:
> open new PR for it and make new constructor with width parameter or setter method so other UIs are not changed / potentially broken

Goldorion commented:
> > open new PR for it and make new constructor with width parameter or setter method so other UIs are not changed / potentially broken

The NumberProcedureSelector was just used by the redstone power level for blocks and the UI was exactly how it should (except it the JSpinner had a bigger width).

KlemenDEV commented:
@MysteriousWolf could you make a new icon for this ME?

AlsoSomeoneElse commented:
Some issues with the "burn duration" tooltips:
- The formula has `20` instead of `200`
- There's an extra space in the French tooltip after `[ici]`

Niketats commented:
so, is it possible to download this as a plugin?

KlemenDEV commented on src/main/java/net/mcreator/element/GeneratableElement.java:
```diff
@@ -39,7 +39,7 @@
 public abstract class GeneratableElement {
 
-	public static final int formatVersion = 29;
+	public static final int formatVersion = 32;
```
32?

Goldorion replied:
I was going to explain it ;)
#2537 seems it will be merged before Wednesday, so I can continue to make changes. Then, #2813 also seems it will be merged, so it adds +2 to the converter number. This PR seems to be the next PR and so, it will require bumping the converter version a few times, but I may not be available for 2 weeks (explained on Discord a few days ago) when it will be required. I'm just taking this advance to not slow down new features in case only the icon is needed for its addition to the master.

Defeatomizer commented on src/main/java/net/mcreator/element/ModElementTypeLoader.java:
```diff
@@ -51,9 +51,7 @@ public static void loadModElements() {
-				new ModElementType<>("fluid", 'f', BaseType.BLOCK, RecipeType.BUCKET, FluidGUI::new, Fluid.class));
-		ModElementType.FUEL = register(
-				new ModElementType<>("fuel", null, BaseType.OTHER, RecipeType.NONE, FuelGUI::new, Fuel.class));
+				new ModElementType<>("fluid", 'u', BaseType.BLOCK, RecipeType.BUCKET, FluidGUI::new, Fluid.class));
```
`u` is used by function MET right below this line, also this is not the PR that adds feature MET. Change this key back.

Defeatomizer commented on src/main/java/net/mcreator/element/converter/fv32/ItemDispenserToItemExtensionConverter.java:
```diff
@@ -0,0 +1,74 @@
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class ItemDispenserToItemExtensionConverter implements IConverter {
```
What is _item dispenser_? I'd name this class `ItemDispenseBehaviorToItemExtensionConverter`.

Defeatomizer commented on src/main/java/net/mcreator/ui/modgui/ItemExtensionGUI.java:
```diff
@@ -0,0 +1,197 @@
+	private final JSpinner layerChance = new JSpinner(new SpinnerNumberModel(0, 0, 1, 0.01));
+	// Dispenser behaviour
+	private final JCheckBox hasDispenseBehavior = L10N.checkbox("elementgui.common.enable");
+	private ProcedureSelector fuelSuccessCondition;
```
Wrong category for this selector.
I also see no point at keeping the comments (I mean, no other GUIs use them but split different parameters categories with empty lines instead).

Defeatomizer commented on src/main/java/net/mcreator/ui/modgui/ItemExtensionGUI.java:
```diff
@@ -0,0 +1,197 @@
+				TitledBorder.DEFAULT_POSITION, getFont(), (Color) UIManager.get("MCreatorLAF.BRIGHT_COLOR")));
+		dispenserBehaviourPanel.setOpaque(false);
+
+		addPage(PanelUtils.totalCenterInPanel(PanelUtils.northAndCenterElement(PanelUtils.join(
```
This would be more readable if you split all this statement into variables with smaller panels.

Defeatomizer commented on src/main/java/net/mcreator/ui/modgui/ItemExtensionGUI.java:
```diff
@@ -0,0 +1,197 @@
+	}
+
+	@Override protected AggregatedValidationResult validatePage(int page) {
+		return new AggregatedValidationResult.PASS();
```
This means you allow those settings to be applied to no item?

Defeatomizer commented on src/main/java/net/mcreator/ui/modgui/ItemExtensionGUI.java:
```diff
@@ -0,0 +1,197 @@
+	}
+
+	@Override public @Nullable URI contextURL() throws URISyntaxException {
+		return new URI(MCreatorApplication.SERVER_DOMAIN + "/wiki/how-make-game-item-extension");
```
You forgot to remove the `game` word when copying the code from `GameRuleGUI`.

KlemenDEV reviewed (commented):
What will happen if there is an item that is used in fuel and at the same time has a dispense behavior? Not sure if current converter will handle this properly

Goldorion commented:
> What will happen if there is an item that is used in fuel and at the same time has a dispense behavior? Not sure if current converter will handle this properly

I already thought this, but I concluded I couldn't do a better thing due to how converters work. As each converter is for one MET only, this means if I convert an item MET inside a fuel converter, MCreator will not know it and it will convert again the custom item to a second Item extension, as we also need the second converter in case a custom item is not a fuel. Another thing is as MCreator just convert one ME one after the other one, I can't check in one of the two converter if the item was also fuel and so, is now an item extension. I would have to do the same code in both converters and it may greatly impact performances in big workspaces.

KlemenDEV commented:
> If you have an idea, I would be happy to hear it because I'm out of ideas currently.

The converter could check if the item extension for the given item already exists (the name of the item extension from the converter is known) and if it does, apply an additional extension. You need to do this for both converters as you can't and shouldn't predict the order of conversions.

Goldorion commented:
> > If you have an idea, I would be happy to hear it because I'm out of ideas currently.
> 
> The converter could check if the item extension for the given item already exists (the name of the item extension from the converter is known) and if it does, apply an additional extension. You need to do this for both converters as you can't and shouldn't predict the order of conversions.

We know the name of the mod element to convert, but we can not know the name of the other mod element. For example, an item ME has already been converted and MCreator wants to convert the fuel ME of the same item. The item extension exists, but its name was the name of the item ME + "Extension". In this case, how can we know the name of the item or item extension ME? The user may have writen a random name.

KlemenDEV commented:
> Edit: I can use a loop to check the item to use for both ME, but this was what I tried to avoid.

This is probably what you will have to do. Indeed names can't be relied upon. You will need to loop MEs, first filtering by type:itemextension and then by item in extension itself.

KlemenDEV commented:
I hope this was well tested

Goldorion commented:
> I hope this was well tested

I made a new test workspace in MCreator 2022.1 with an item and a fuel (+ a condition for the dispenser behaviour) and tested it 2-3 times. Then, I took the backup, deleted the item mod element and created a new one (and changed the fuel ME of course), to have a fuel and then, an item and re-tested 2-3 times and every time, everything worked. I don't know what I can do to test more.

AlsoSomeoneElse commented on src/main/java/net/mcreator/element/converter/fv31/ItemDispenseBehaviorToItemExtensionConverter.java:
```diff
@@ -0,0 +1,94 @@
+						if (item.get("dispenseResultItemstack") != null)
+							itemExtension.dispenseResultItemstack = new Procedure(
+									item.get("dispenseResultItemstack").getAsJsonObject().get("name").getAsString());
+						return itemExtension;
```
Wouldn't this convert the item element into an item extension element?

Goldorion replied:
When I tested, the item mod element always stayed in the workspace, but I think I understand what you mean. I will change it and test to see if it affects something.

Goldorion replied:
It does not seem to affect the converter as the item and item extension MEs can still be open properly, but it is surely better to return the item ME.

Goldorion commented on src/main/java/net/mcreator/element/converter/fv31/ItemDispenseBehaviorToItemExtensionConverter.java:
```diff
@@ -0,0 +1,94 @@
+import net.mcreator.element.ModElementType;
+import net.mcreator.element.converter.IConverter;
+import net.mcreator.element.parts.MItemBlock;
+import net.mcreator.element.parts.NumberProcedure;
```
Oopsi

KlemenDEV commented:
>  I don't know what I can do to test more.

Nothing, just asked if it was tested and for sure it was:)

KlemenDEV commented:
I did some changes on the UI to make it look more MCreator-ish.

@AlsoSomeoneElse, would you be ready to take this PR and do a final review and merge if OK?

AlsoSomeoneElse commented on plugins/generator-1.18.2/forge-1.18.2/itemextension.definition.yaml:
```diff
@@ -0,0 +1,18 @@
+    writer:  java
+    conditon: hasDispenseBehavior
+    deleteWhenConditionFalse: true
+    name: "@SRCROOT/@BASEPACKAGEPATH/item/extension/@JavaModNameItemExtension.java"
```
Should be `@NAMEItemExtension.java`

AlsoSomeoneElse commented on src/main/java/net/mcreator/element/converter/fv32/FuelToItemExtensionConverter.java:
```diff
@@ -0,0 +1,92 @@
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+/*
```
Duplicated copyright notice, same for the other converter class

Defeatomizer commented on plugins/generator-1.18.2/forge-1.18.2/templates/dispensebehaviour.java.ftl:
```diff
@@ -0,0 +1,68 @@
+
+package ${package}.item.extension;
+
+@Mod.EventBusSubscriber(bus = Mod.EventBusSubscriber.Bus.MOD) public class ${JavaModName}ItemExtension {
```
```suggestion
@Mod.EventBusSubscriber(bus = Mod.EventBusSubscriber.Bus.MOD) public class ${name}ItemExtension {
```
I'd also call the class `name + "DispenseBehavior"` because other parameters of item extension are not handled in this class for MC 1.18.2.
